adapters/mqtt-vertx-base/src/main/java/org/eclipse/hono/adapter/mqtt/AbstractVertxBasedMqttProtocolAdapter.java

```diff
@@ -730,47 +731,51 @@ protected final void onUnsubscribe(
-                        if (failure instanceof ClientErrorException) {
-                            // issue credit so that application(s) can send the next command
-                            commandContext.reject(getErrorCondition(failure), 1);
-                        } else {
-                            commandContext.release(1);
-                        }
-                        metrics.reportCommand(
-                                command.isOneWay() ? Direction.ONE_WAY : Direction.REQUEST,
-                                sub.getTenant(),
-                                tenantTracker.result(),
-                                ProcessingOutcome.from(failure),
-                                command.getPayloadSize(),
-                                timer);
-                        return null;
-                    });
-                },
-                remoteClose -> {},
-                livenessCheckInterval);
+        final Handler<CommandContext> commandHandler = commandContext -> {
+
+            Tags.COMPONENT.set(commandContext.getCurrentSpan(), getTypeName());
+            final Sample timer = metrics.startTimer();
+            final Command command = commandContext.getCommand();
+            final Future<TenantObject> tenantTracker = getTenantConfiguration(sub.getTenant(),
+                    commandContext.getTracingContext());
+
+            tenantTracker.compose(tenantObject -> {
+                if (!command.isValid()) {
+                    return Future.failedFuture(new ClientErrorException(HttpURLConnection.HTTP_BAD_REQUEST,
+                            "malformed command message"));
+                }
+                return checkMessageLimit(tenantObject, command.getPayloadSize());
+            }).compose(success -> {
+                addMicrometerSample(commandContext, timer);
+                onCommandReceived(tenantTracker.result(), mqttEndpoint, sub, commandContext,
+                        cmdHandler);
+                return Future.succeededFuture();
+            }).otherwise(failure -> {
+                if (failure instanceof ClientErrorException) {
+                    // issue credit so that application(s) can send the next command
+                    commandContext.reject(getErrorCondition(failure), 1);
+                } else {
+                    commandContext.release(1);
+                }
+                metrics.reportCommand(
+                        command.isOneWay() ? Direction.ONE_WAY : Direction.REQUEST,
+                        sub.getTenant(),
+                        tenantTracker.result(),
+                        ProcessingOutcome.from(failure),
+                        command.getPayloadSize(),
+                        timer);
+                return null;
+            });
+        };
+        final Handler<Void> remoteCloseHandler = remoteClose -> {};
+        if (sub.getAuthenticatedDeviceId() != null && !sub.getAuthenticatedDeviceId().equals(sub.getDeviceId())) {
```

Comment:
this check looks like it should be implemented in a method (e.g. *isGatewaySubscription*) in `Subscription`, don't you think?

Reply:
Good point, I've refactored it accordingly.